include/drivers/video.h

```diff
@@ -0,0 +1,556 @@
+ * event(s) is(are) generated for discarded buffers, and driver calls registered
+ * event callback (if any). The user has to dequeue the remaining buffers with
+ * video_dequeue().
+ *
+ * @retval 0 Is successful.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_stream_stop(struct device *dev)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	__ASSERT(api->stream_stop,
+		 "stream_stop must be implemented by driver");
+
+	return api->stream_stop(dev);
+}
+
+/**
+ * @brief Get the capabilities of a video endpoint.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param caps Pointer to the video_caps struct to fill.
+ *
+ * @retval 0 Is successful, -ERRNO code otherwise.
+ */
+static inline int video_get_caps(struct device *dev,
+				 enum video_endpoint_id ep,
+				 struct video_caps *caps)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	__ASSERT(api->get_caps, "get_caps must be implemented by driver");
+
+	return api->get_caps(dev, ep, caps);
+}
+
+/**
+ * @brief Set the value of a control.
+ *
+ * This set the value of a video control, value type depends on control ID, and
+ * must be interpreted accordingly.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cid Control ID.
+ * @param value Pointer to the control value.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -ENOTSUP If format is not supported.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_set_ctrl(struct device *dev, unsigned int cid,
+				 void *value)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->set_ctrl == NULL) {
+		return -ENOTSUP;
+	}
+
+	return api->set_ctrl(dev, cid, value);
+}
+
+/**
+ * @brief Get the current value of a control.
+ *
+ * This retrieve the value of a video control, value type depends on control ID,
+ * and must be interpreted accordingly.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cid Control ID.
+ * @param value Pointer to the control value.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -ENOTSUP If format is not supported.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_get_ctrl(struct device *dev, unsigned int cid,
+				 void *value)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->get_ctrl == NULL) {
+		return -ENOTSUP;
+	}
+
+	return api->get_ctrl(dev, cid, value);
+}
+
+/**
+ * @brief Register/Unregister event callback for a video endpoint.
+ *
+ * Register a endpoint event handler which is called on frame completion,
+ * error, etc (cf video_event enum). There is no guarentee on the
```

Comment:
So this answers my above comments about the context of the callback. If there is no guarantee of context the caller must assume an IRQ context which is the least useful place to get a callback as it effectively means your going to use some form of IPC or task queue to do the actual work. In which case, why not use a k_poll-able signal here instead

Reply:
Yes I tend to agree here, what do you think about replacing register_event_callback with register_event_signal, this would allow the user to poll on the signal, and use standard dequeue function on e.g. BUFFER_DONE event.

Reply:
I think that would be fine, the exception might be that you'd like to actually have code run in an ISR context. For example if it really can't be scheduled for later then using a signal would not make sense, but this would need to be clearer about that.

Reply:
Yes, the only code we would like to run in IRQ context is for the video device management, and it's already done in the driver anyway, and there is no guarantee for the callback context (it depends on the driver). Most of the time, the event will lead to dequeue a buffer and process it, or requeue it to another video device. The only problem I see with k_poll is that we can miss events in case signals come faster than managed in user thread, which then needs to be taken into account.

Reply:
I've added video_set_signal() allowing to register an asynchonous signal which is then signaled on video endpoint buffer event.